### PR TITLE
feat: use same LLVM across development tooling

### PR DESCRIPTION
**Pull request by tomberek**
## Proposed Changes
- Ensure we use the same version of llvm/clang/etc throughout development.
- adds ci devShell (so far not used)

Are the clang-14 or clang-16 versions required or necessary for something?

Alternative: override llvmPackages_* and llvm_* and clang-tools_* in the overlay; that is more comprehensive, but less targeted/precise.

Note: overall bump to release-23.11 would make clang_16 the default. release-23.05 uses 11.1 by default.

## Release Notes
N/A

### CODE REVIEW

File: flake.nix [hunk 1]
```diff
@@ -159,7 +159,7 @@
           rust.cargoManifestPath = "cli/Cargo.toml";
         };
         tools = {
-          clang-tools = final.clang-tools_16;
```
Comment (ysndr) [inline flake.nix:162]: why downgrade to 15?
we use some clang-format rules only available on >=16.

**Review by ysndr — CHANGES_REQUESTED** [review]
if all of the llvm packages are available in version 16, maybe we should stick with that.
Except some formatting rules I don't know whether we _required_ 16, and the rules that did might not be entirely essential 

See also: https://github.com/flox/flox/pull/670#discussion_r1431713089

File: flake.nix (hunk 1 above)
Comment (tomberek) [reply flake.nix:162]: I tried 16 and had failures for both ccls and include-what-you-use that didnt seem to have quick fixes. Upgrading to a newer nixpkgs may help here, but that is a separate PR.

Comment (ysndr) [reply flake.nix:162]: Can you check that clang-format is still formatting?
I had some issues with it in the past where it jsut didn't format if it couldn't parse the config file

Comment (ysndr) [reply flake.nix:162]: I guess we can just set `breakbeforeconceptdeclarations` to true and downgrade to 15 through. the other setting we were discussing in the comment i linked was added in clang-format 15 actually

Comment (tomberek) [reply flake.nix:162]: Yes, it still functions as far as I can tell. Applying clang-format to all of ./pkgdb produces a few changes. Another option here is to keep clang-tools where it is.

File: pkgs/flox-pkgdb/default.nix [hunk 2]
```diff
@@ -27,13 +28,15 @@
...
-  llvm, # for `llvm-symbolizer'
+  llvm_15, # for `llvm-symbolizer'
   gdb ? throw "`gdb' is required for debugging with `g++'",
   lldb ? throw "`lldb' is required for debugging with `clang++'",
   valgrind ? throw "`valgrind' is required for memory sanitization on Linux",
   substituteAll,
   symlinkJoin,
 }: let
+  llvmPackages = llvmPackages_15;
+  llvm = llvm_15;
```
Comment (ysndr) [inline pkgs/flox-pkgdb/default.nix:39]: isnt llvm_15 not just llvmPackages_15.llvm?

File: pkgs/flox-pkgdb/default.nix [hunk 3]
```diff
@@ -185,11 +188,15 @@ in
...
             # For IDEs
-            ccls
+            (ccls.override (old: {
+              llvmPackages = llvmPackages;
+            }))
             bear
             # For lints/fmt
-            clang-tools_16
-            include-what-you-use
+            clang-tools_15
```
Comment (ysndr) [inline pkgs/flox-pkgdb/default.nix:196]:
```suggestion
            clang-tools_16
```

File: pkgs/flox-pkgdb/default.nix [hunk 4]
```diff
@@ -5,8 +5,9 @@
   doxygen,
   bear,
   boost,
+  llvmPackages_15,
   ccls,
-  clang-tools_16,
+  clang-tools_15,
```
Comment (ysndr) [inline pkgs/flox-pkgdb/default.nix:10]:
```suggestion
  clang-tools_16,
```

**Pull request closed without merging**